design/kubernetes-friendly-oam.md

```diff
@@ -0,0 +1,482 @@
+but it may be omitted when authoring an ApplicationConfiguration. Metadata such
+as names and labels may be filled automatically by the OAM runtime if a distinct
+instance of a trait configuration is created as part of processing an
+ApplicationConfiguration. As with workloads, Kubernetes runtimes may treat
+traits as an [embedded object]. Each trait resource is embedded beneath the key
+`trait`. This allows the spec to add other keys alongside `trait` in future
+without breaking compatibility.
+
+### Scopes
+
+This document proposes that application scopes, while maintaining their current
+purpose of grouping application components, function identically to traits from
+an API perspective. Available kinds of scope are defined via a ScopeDefinition,
+which also specifies whether a component may exist in multiple scopes of a
+particular kind:
+
+```yaml
+apiVersion: core.oam.dev/v1alpha1
+kind: ScopeDefinition
+metadata:
+  name: networkscope.core.oam.dev
+spec:
+  allowComponentOverlap: true
+  crdName: networkscope.core.oam.dev
+```
+
+The above ScopeDefinition references a Kubernetes CustomResourceDefinition named
+networkscope.core.oam.dev, designating custom resources defined by said CRD to
+be valid OAM scopes. As with the other OAM definition resources OAM runtimes
+that are not powered by Kubernetes could opt to either implement the CRD
+concept, or introduce a new field to the ScopeDefinition spec that defines a
+kind of scope and its schema. Application operators can discover available OAM
+scope kinds by running `kubectl get scopedefinitions`. Scopes would most
+frequently be defined by OAM runtimes or infrastructure operators.
+
+```yaml
+apiVersion: core.oam.dev/v1alpha1
+kind: NetworkScope
+metadata:
+  name: example-vpc-network
+  labels:
+    region: us-west
+    environment: production
+spec:
+  networkId: cool-vpc-network
+  subnetIds:
+  - cool-subnetwork
+  - cooler-subnetwork
+  - coolest-subnetwork
+  internetGatewayType: nat
+```
+
+Scopes would be configured by creating an instance of configuration - a scope
+resource that may be referenced by ApplicationConfigurations. This is not
+functionally dissimilar to the contemporary OAM specification, which uses an
+ApplicationConfiguration to create an instance of scope configuration that is
+referenced by other ApplicationConfigurations.
+
+### Application Configuration
+
+Finally, the ApplicationConfiguration resource ties together a set of
+components, traits, and scopes as it does today:
+
+```yaml
+apiVersion: core.oam.dev/v1alpha1
+kind: ApplicationConfiguration
+metadata:
+  name: cool-example
+spec:
+  components:
+  - componentName: example-server
+    parameterValues:
+    - name: instanceName
+      value: cool-example
+    - name: cacheSecret
+      value: cache-connection
+    traits:
+      trait:
+      - apiVersion: core.oam.dev/v1alpha1
+        kind: ManualScalerTrait
+        spec:
+          replicaCount: 3
```

Comment:
I think the format here should change like this:

```
    traits:
      - trait:
           apiVersion: core.oam.dev/v1alpha1
           kind: ManualScalerTrait
           spec:
             replicaCount: 3
```

This is like scopeRef

Reply:
After deep consideration, I think we'd better use `name` and `version` in definitionRef.

```
    traits:
      - name: manualscalertrait.core.oam.dev
         version: v1alpha1
         schematic:
           replicaCount: 3
```

I have these reasons:

1. Same experience with traitDefinition. Users could easily learn how to write, not bother learn schema twice.
2. In the OAM Spec, we just need to leave `schematic` to be custom defined. This could not only be more clean, but also be easier for non-k8s implementation to define the schematic in traitDefinition.
3. The way we validate the trait in application configuration will be more intuitive for both K8s and non-K8s implementation. We follow the schema from trait definition (in fact it's definitionRef), then we can index the schema directly from definitionRef with name and version. If you use apiVersion and kind here, just think for non-k8s implementation, they have to implement same mechanism like this.

Reply:
This is not align with what we discussed in the meeting. We agreed to use K8s resource model for end user facing objects (specially, for trait spec, we will recommend this model). The intention is to eliminate confusion and create a "k8s-like" resource here which is not friendly to either K8s or non-K8s implementation.

Reply:
> I think the format here should change like this

Good call! What you propose in your first comment is what I had intended to write; i.e. the `traits` field should contain an array. I'll update.